src/codegen/solana_deploy.rs

```diff
@@ -314,13 +302,57 @@ pub(super) fn solana_deploy(
         );
 
         // Calculate minimum balance for rent-exempt
-        let (space, lamports) = if let Some(ConstructorAnnotation::Space(space_expr)) = func
-            .annotations
-            .iter()
-            .find(|tag| matches!(tag, ConstructorAnnotation::Space(..)))
-        {
-            let space_var = vartab.temp_name("space", &Type::Uint(64));
+        let (space, lamports) = if let Some((_, space_expr)) = &func.new_annotations.space {
             let expr = expression(space_expr, cfg, contract_no, None, ns, vartab, opt);
+            // If the space has not been resolved into a literal in sema,
+            // we must verify if we are allocating enough space during runtime.
+            if !matches!(expr, Expression::NumberLiteral { .. }) {
```

Comment:
so what about when it's a constant expression?

Reply:
If look carefully at `src/sema/function_annotation.rs`, between there can be only four options when resolving an annotation:
1. It is a repeated one, in which case we error out.
2. It is a literal, in which case we save it.
3. It is resolved by `eval_const_number`, which returns a `BigInt`. We save this big it in a number literal.
4. `eval_const_number` errors out, in which case we discard the annotation and never enter codegen.

By the paths no. 2 and no. 3, I am faithfully assuming we are not going to have constant expressions in codegen.

Reply:
The ast should be a representation of the source code, and not have constant expressions folded.

Reply:
Fixed